pylabrobot/pumps/agrowpumps/agrowdosepump.py

```diff
@@ -0,0 +1,185 @@
+
+  async def run_continuously(self, speed: List[float],
+                             use_channels: List[int]):
+    """Run pumps at the specified speeds.
+        Args:
+            speed: rate at which to run pump.
+            use_channels: pump array channels to run
+
+        Raises:
+            ValueError: Pump address out of range
+            ValueError: Pump speed out of range
+    """
+    assert self.modbus is not None, "Modbus connection not established"
+    assert self.pump_index_to_address is not None, "Pump address mapping not established"
+    if any(channel not in range(1, self.num_channels + 1) for channel in use_channels):
+      raise ValueError(f"Pump address out of range for this pump array. \
+        Value should be between 1 and {self.num_channels+1}")
+    for pump_index, pump_speed in zip(use_channels, speed):
+      pump_speed = int(pump_speed)
+      if pump_speed not in range(101):
+        raise ValueError("Pump speed out of range. Value should be between 0 and 100.")
+      self.modbus.write_register(self.pump_index_to_address[pump_index], pump_speed, unit=self.unit)
+
+  async def halt(self):
+    """ Halt the entire pump array. """
+    assert self.modbus is not None, "Modbus connection not established"
+    assert self.pump_index_to_address is not None, "Pump address mapping not established"
+    print("Engaging shutdown routine")
+    for pump in self.pump_index_to_address:
+      address = self.pump_index_to_address[pump]
+      self.modbus.write_register(address, 0, unit=self.unit)
+
+  async def stop(self):
+    """ Close the connection to the pump array. """
+    await self.halt()
+    assert self.modbus is not None, "Modbus connection not established"
+    self.modbus.close()
+    assert self.keep_alive_thread is not None, "Keep alive thread not established"
+    self.keep_alive_thread.join()
+
+
+class SimulatedModbusClient:
+  """
+  SimulatedModbusClient is a class that allows users to simulate Modbus communication.
+
+  Attributes:
+      connected: A boolean that indicates whether the simulated client is connected.
+
+  """
+
+  def __init__(self, connected: bool = False):
+    self.connected = connected
+
+  async def connect(self):
+    self.connected = True
+
+  def read_holding_registers(self, *args, **kwargs):
+    assert self.connected, "Modbus connection not established"
+    for arg in kwargs.items():
+      if arg == "unit":
+        continue
+    if args[0] == 19:
+      return [16708, 13824, 0, 0, 0, 0, 0][:args[1]]
+    if args[0] == 0:
+      return
+
+  def write_register(self, *args, **kwargs):
+    assert self.connected, "Modbus connection not established"
+    if "unit" not in kwargs:
+      raise ValueError("unit must be specified")
+    if args[0] in range(100, 107):
+      return
+    else:
+      raise ValueError("address out of range")
+
+  def close(self):
+    assert self.connected, "Modbus connection not established"
+    self.connected = False
+    del self
```

Comment:
Why should closing the connection destroy the object? I think more likely than not it's fine to have Python destroy the object when it's no longer referenced.

Reply:
I'll delete line 185